### ISSUE TITLE
Error when fitting with zero fraction = 75%

### ISSUE THREAD

**Issue by fperraudeau** [drisso/archive-zinbwave#30]
require(zinb)

When loading zinb, I get the following warning messages:
_Loading required package: zinb
Warning: namespace ‘bigmemory’ is not available and has been replaced
by .GlobalEnv when processing object ‘plot.index’
Warning: namespace ‘bigmemory’ is not available and has been replaced
by .GlobalEnv when processing object ‘plot.index’_

Load counts (just for me to remember counts = t(simData[[3]]$counts) in sim_allen_75.rda).

     load('datasets/sim_allen75_error1.rda') 

Check that no gene with zero count.

     sum(rowSums(counts) == 0)

Can we set a seed for zinbFit? If yes, I did not find where to set it. 

    set.seed(98257)
    zinbFit(counts, K = 4) 

I get the following error: _"Error in if (abs((total.lik[iter] - total.lik[iter - 1])/total.lik[iter -  : \n  missing value where TRUE/FALSE needed\n"._ Now, if I call zinbFit a second time

    zinbFit(counts, K = 4)

I get the following error: _"Error in optim(fn = zinb.loglik.regression, gr = zinb.loglik.regression.gradient,  :  initial value in 'vmmin' is not finite"._ Now, if I call zinbFit a third time

    set.seed(732)
    zinbFit(counts, K = 4)
    
No error!  

With the previous commands, I also get warnings, but I think last time you said the warnings are ok.
_In addition: Warning messages:
1: In simpute.als(x, J, thresh, lambda, maxit, trace.it, warm.start,  :
  Convergence not achieved by 100 iterations
2: In optimize(f = zinb.loglik.dispersion, Y = as.matrix(Y), mu = mu,  :
  NA/Inf replaced by maximum positive value_

**Comment by fperraudeau** [comment]
@jpvert, just an update.

I now simulate the datasets using the method we talked about this week, it is much better and flexible. But, I still have the errors described above when I call _zinbFit()_.

A bientot,
Fanny

**Comment by jpvert** [comment]
Ok I found the problem. Basically, it is dangerous to fit a number of parameters for a gene which is larger than the number of non-zero counts for the gene. In the above example, there are genes with only 3 non-zero counts

    > min(rowSums(counts))
    [1] 3

but we fit K=4 latent parameters plus a gene-specific offset, meaning we fit a total of 5 parameters for each gene. For a gene with 3 non-zero counts, 3 parameters are enough to fit very well the non-zero counts (i.e., having a small probability of dropout, and a mean value close to the observed counts). Then there remains 2 degrees of freedom to fit all other zeros, which is basically obtained by having a small mu / small p, or a large mu / large p. The likelihood is then very flat on the 2-dimensional null space, and I observe that the log-mu / logit-pi values tend to get pretty large; the error then comes from the fact that when they get too large/small, we get a +Inf or -Inf in the likelihood.

I see 3 options to fix this:
- rewrite the computation of log-likelihood and gradients, to work more carefully in the log space. This is doable but requires some efforts (not use the C-coded distributions functions of R anymore). This would remove the error, but not the fact that the coefficients could get arbitrarily large, which is probably not what we want
- Enforce that the minimum number of non-zero counts per row and column is always at least K+1 (or K if there is no offset). This would remove the null subspace where the likelihood is almost constant. In our case, for K=4 this means removing genes (and cells) with less than 5 counts. This seems to make sense.
- Increase regularization through epsilon. This is why we have regularization, to prevent the parameters to get too large. The default value is pretty small (1e-6), and not enough to prevent the coefficients to get "large" (say in the 100's). Just setting epsilon=1 fixes the problem in the above example.

    > zinbFit(counts, K = 4, ncores=6, epsilon=1) 
    Object of class ZinbModel.
    100 samples; 1000 genes.
    1 sample-level covariates; 1 gene-level covariates; 4 latent factors.

In conclusion, I'd suggest to make no change in the code but simply mention that an error is possible when there are genes with less than K+1 nonzero counts; if it is important to keep all genes, just increase lambda.

**Comment by fperraudeau** [comment]
Great !

Indeed, when I filter out genes with less than 5 counts, I don't have errors. But, the clustering is bad, i.e., cells from different clusters are not well separated.

When I filter genes with only zero counts (so keep genes with less than 5 counts), but change epsilon to 1000, errors are removed and clustering is great for datasets simulated from Allen. See figure below with epsilon = {1,10,100,1000}. I now set epsilon = number of genes, and we will see if it scales when we change the number of genes/cells. 

Conclusion: the problem seems to be solved. Thank you Jean-Philippe. And, let's keep an eye the epsilon.

<img width="1182" alt="screen shot 2017-01-26 at 14 00 21" src="https://cloud.githubusercontent.com/assets/8387277/22332178/d30e6612-e3cf-11e6-9f9e-58b1c6bb2c96.png">

**Comment by drisso** [comment]
It seems that I can close this issue
